### PR TITLE
Access Control: Pass db session to hooks

### PR DESCRIPTION
**Pull request by kalleep**
**What this PR does / why we need it**:
Move the hook call to db layer of resource permissions allow us to pass the transaction session. That way we can ensure that the set operation would roll back if the hook fails.

Also change the signature of resource permissions service to take "friendly named" permission and translate them to actions internally.

Will be used by https://github.com/grafana/grafana/pull/43991?notification_referrer_id=NT_kwDOAWRi1bMyOTQzMDIyMjYzOjIzMzU2MTE3#pullrequestreview-862283996

### CODE REVIEW

File: pkg/services/accesscontrol/models.go [hunk 1]
```diff
@@ -237,6 +237,7 @@ type SetResourcePermissionCommand struct {
 	Actions    []string
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/models.go:237]: nit: do we still need `Actions` as part of this command?

**Review by IevaVasiljeva — APPROVED** [review]
This looks really good! And I'm glad that the changes seem to have made the code simpler rather than more complicated, amazing work!

File: pkg/services/accesscontrol/database/resource_permissions.go [hunk 2]
```diff
@@ -32,7 +32,11 @@ func (p *flatResourcePermission) Managed() bool {
 	return strings.HasPrefix(p.RoleName, "managed:")
 }
 
-func (s *AccessControlStore) SetUserResourcePermission(ctx context.Context, orgID, userID int64, cmd accesscontrol.SetResourcePermissionCommand) (*accesscontrol.ResourcePermission, error) {
+func (s *AccessControlStore) SetUserResourcePermission(
+	ctx context.Context, orgID, userID int64,
+	cmd accesscontrol.SetResourcePermissionCommand,
+	hook func(session *sqlstore.DBSession, orgID, userID int64, resourceID, permission string) error,
```
Comment (Jguer) [inline pkg/services/accesscontrol/database/resource_permissions.go:38]: nit: could we get a type alias for the hook func to improve readability and cross reference?

File: pkg/services/accesscontrol/database/resource_permissions.go [hunk 3]
```diff
@@ -87,6 +105,12 @@ func (s *AccessControlStore) SetBuiltInResourcePermission(ctx context.Context, o
 
 	err = s.sql.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
 		permission, err = s.setResourcePermission(sess, orgID, managedBuiltInRoleName(builtInRole), s.builtInRoleAdder(sess, orgID, builtInRole), cmd)
+		if err != nil {
+			return err
+		}
+		if hook != nil {
+			return hook(sess, orgID, builtInRole, cmd.ResourceID, cmd.Permission)
+		}
 		return err
```
Comment (Jguer) [inline pkg/services/accesscontrol/database/resource_permissions.go:114]: shouldn't this be specifically return nil?

```suggestion
		return nil
```

Comment (Jguer) [reply pkg/services/accesscontrol/database/resource_permissions.go:114]: or 

```go
		if err == nil && hook != nil {
			return hook(sess, orgID, builtInRole, cmd.ResourceID, cmd.Permission)
		}
		
		return err
```

Comment (Jguer) [reply pkg/services/accesscontrol/database/resource_permissions.go:114]: should match L78-84 at least :)

**Pull request merged**